testing/mock/_asserts.ts

```diff
@@ -0,0 +1,324 @@
+  // deno-lint-ignore no-explicit-any
+  Args extends unknown[] = any[],
+  // deno-lint-ignore no-explicit-any
+  Return = any,
+> {
+  /** Arguments passed to a function when called. */
+  args?: [...Args, ...unknown[]];
+  /** The instance that a method was called on. */
+  self?: Self;
+  /**
+   * The value that was returned by a function.
+   * If you expect a promise to reject, expect error instead.
+   */
+  returned?: Return;
+  error?: {
+    /** The class for the error that was thrown by a function. */
+    // deno-lint-ignore no-explicit-any
+    Class?: new (...args: any[]) => Error;
+    /** Part of the message for the error that was thrown by a function. */
+    msgIncludes?: string;
+  };
+}
+
+/**
+ * Asserts that a spy is called as expected.
+ * Returns the call.
+ */
+export function assertSpyCall<
+  Self,
+  Args extends unknown[],
+  Return,
+>(
+  spy: Spy<Self, Args, Return>,
+  callIndex: number,
+  expected?: ExpectedSpyCall<Self, Args, Return>,
+) {
+  if (spy.calls.length < (callIndex + 1)) {
+    throw new AssertionError("spy not called as much as expected");
+  }
+  const call: SpyCall = spy.calls[callIndex];
+  if (expected) {
+    if (expected.args) {
+      try {
+        assertEquals(call.args, expected.args);
+      } catch (e) {
+        assertIsError(e);
+        throw new AssertionError(
+          "spy not called with expected args:\n" +
+            e.message.split("\n").slice(1).join("\n"),
+        );
+      }
+    }
+
+    if ("self" in expected) {
+      try {
+        assertEquals(call.self, expected.self);
+      } catch (e) {
+        assertIsError(e);
+        let message = expected.self
+          ? "spy not called as method on expected self:\n"
+          : "spy not expected to be called as method on object:\n";
+        message += e.message.split("\n").slice(1).join("\n");
+        throw new AssertionError(message);
+      }
+    }
+
+    if ("returned" in expected) {
+      if ("error" in expected) {
+        throw new TypeError(
+          "do not expect error and return, only one should be expected",
+        );
+      }
+      if (call.error) {
+        throw new AssertionError(
+          "spy call did not return expected value, an error was thrown.",
+        );
+      }
+      try {
+        assertEquals(call.returned, expected.returned);
+      } catch (e) {
+        assertIsError(e);
+        throw new AssertionError(
+          "spy call did not return expected value:\n" +
+            e.message.split("\n").slice(1).join("\n"),
+        );
+      }
+    }
+
+    if ("error" in expected) {
+      if ("returned" in call) {
+        throw new AssertionError(
+          "spy call did not throw an error, a value was returned.",
+        );
+      }
+      assertIsError(
+        call.error,
+        expected.error?.Class,
+        expected.error?.msgIncludes,
+      );
+    }
+  }
+  return call;
+}
+
+/**
+ * Asserts that an async spy is called as expected.
+ * Returns the call.
```

Comment:
I think returning value from assertion is confusing and less readable

(Note: We had similar discussion in the past about this design https://github.com/denoland/deno_std/pull/1052#discussion_r674510375